### DISCUSSION TITLE
Issue implementing Neo Hookean material model, unable to converge

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#28622] [Q&A Modules: Solid mechanics] [answered]
Hi, I am following the instructions to implement a custom hyperelastic model by inheriting from ComputeLagrangianStressPK1.

I have a material file, `ComputeStrainEnergyNeoHookean`  where I calculated the strain energy, W. In that material file, I also calculated and declared the derivative of W wrt F, the deformation gradient (which is the PK1 stress), and the second derivative (which is the tangent operator):

```
	_strain_energy(declareProperty<Real>(_base_name + "strain_energy")),
	_dWdF(declarePropertyDerivative<RankTwoTensor>(_base_name + "strain_energy", _base_name + "deformation_gradient")),	// this is the stress
	_d2WdF2(declarePropertyDerivative<RankFourTensor>(_base_name + "strain_energy", _base_name + "deformation_gradient", _base_name + "deformation_gradient")) //this is the tangent operator
```

I have an "intermediate" file, which I called `NeoHookeanStressIntermediate`, to get those derivatives, and declare them as material properties. I did this because I realized that `ComputeLagrangianStressBase` is derived from `Material` instead of `DerivativeMaterialInterface`, so it could not recognize the command `getMaterialPropertyDerivative` in my custom constructor. 

Finally, I have the file for calculating the stress, i.e `ComputeStressNeoHookean`. I overrode the method `computeQpPK1Stress()`, and within that method, I set 

```
	_pk1_stress[_qp] = _dWdF_in[_qp];
	
	_pk1_jacobian[_qp] = _d2WdF2_in[_qp];
```

where dWdF and d2WdF2 are the derivatives I calculated from my Material file, and declared with my 'intermediate' file.

I then set up my input file, and tried to run the simulation, but I get the error:

```
LEGACY MODES ENABLED:
 This application uses the legacy initial residual evaluation behavior. The legacy behavior performs an often times redundant residual evaluation before the solution modifying objects are executed prior to the initial (0th nonlinear iteration) residual evaluation. The new behavior skips that redundant residual evaluation unless the parameter Executioner/use_pre_smo_residual is set to true. To remove this message and enable the new behavior, set the parameter 'use_legacy_initial_residual_evaluation_behavior' to false in *App.C. Some tests that rely on the side effects of the legacy behavior may fail/diff and should be re-golded.


Postprocessor Values:
+----------------+----------------+
| time           | displace_x     |
+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |
+----------------+----------------+

Pre-SMO residual: 8.24621e-05

Performing automatic scaling calculation

 0 Nonlinear |R| = 8.246211e-05
    Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                   PC failed due to FACTOR_NUMERIC_ZEROPIVOT
  Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
 Solve Did NOT Converge!
Currently Executing
  Finished Solving                                                                       [  5.58 s] [  208 MB]
Aborting as solve did not converge
```

As for the legacy residual stuff, I'm guessing it has to do with my app? I created it in 2023 and recently updated MOOSE. How can I update my app?

But that said, I'm guessing the FACTOR_NUMERIC_ZEROPIVOT comes from the material properties (stress and tangent operators, maybe?).  

Also, how can I enforce the incompressibility constraint?

The input file is here: https://github.com/richmondodufisan/purple/tree/main/Cornea_Hyperelastic_NeoHookean

And the Material files are in my app's folders

**Comment by GiudGiud** [comment]
Hello

For the legacy stuff you can see this message
https://mooseframework.inl.gov/newsletter/2024/2024_04.html

For the convergence issues, you can try adding a `pc_factor_shift` of NONZERO to the petsc options to see if that helps.
If it is truly a problem with the material properties, it likely wont be enough.
You can get more information on how many DOFs are affected using the svd monitor on a small problem.
See this page for more troubleshooting help
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

I would also output all the material properties to exodus and first look for those that are zero when they should not be, then try to match them by hand calculations

**Reply by richmondodufisan** [reply]
Thank you, adding "-pc_factor_shift_type" NONZERO worked.

The second issue I mentioned was enforcing the incompressibility constraint. Now, instead of just defining 

PK1_stress = dW/dF,

I want to add the incompressibility constraint with a Lagrange Multiplier by:

PK1_stress = dW/dF - p*F.inv()

EDIT: The kernel enforcing the constraint should be (J - 1) * del_p where del_p is the variation of the pressure. But my solution isn't converging.

**Reply by GiudGiud** [reply]
If you are just trying to add a new term to the equation you can definitely do that with an extra kernel. 

I don't know enough to help further on this topic

**Reply by richmondodufisan** [reply]
No worries, thank you for your help still. How can I export all material properties to exodus?

**Reply by GiudGiud** [reply]
In every material block you can specify which properties to output (or use 'all') and which output for them (or use 'all')

**Reply by richmondodufisan** [reply]
I have figured out the problem, but I am not sure how to solve it. For the first implementation I did before enforcing compressibility, I was getting results for any stretch ratio. However, when I set the view type to wireframe, I found that only one element was being stretched.

This kind of makes sense as the strain energy was defined as a function of the deformation gradient, and so basically what I assume happened is that the deformation gradient was only non-zero where the boundary condition was applied:

![img1](https://github.com/user-attachments/assets/7eb7a268-9202-46f9-ba2a-73b3cacfd84d)

Then I implemented the incompressible case, and now the volume is not allowed to change, so it only converged for very, very small applied displacements:

![img2](https://github.com/user-attachments/assets/6789c189-a093-4266-98ef-d0a6282287e9)

I'm not sure if I'm not retrieving the deformation gradient properly or if it is something else. I thought maybe I could try using an UpdatedLagrangianFormulation and also running it quasistatically (so it would update as I went along), but for some reason after changing to UpdatedLagrangianStressDivergence it keeps giving an error to add "use_displaceed_mesh = true" even though I already added it! It's like it can't see it in my input file for some reason.

So I switched back to TotalLagrangian, and still tried to run it quasistatically. But it still didn't deform beyond the first element.

Was I wrong to define my strain energy based off of "deformation_gradient"?

@hugary1995 @dschwen mentioning you because this thread is a follow up to https://github.com/idaholab/moose/discussions/27292

The mentioned file is test_2.i

https://github.com/richmondodufisan/purple/tree/main/Cornea_Hyperelastic_NeoHookean

**Reply by richmondodufisan** [reply]
For the incompressible Neo-Hookean model implemented, these are the details below:

First, the potential energy functional is defined as:

![neohook3](https://github.com/user-attachments/assets/5fb11820-10df-4c85-ac57-16fcd9ee5e27)

with p, the pressure being an arbitrary lagrangian multiplier to enforce J = 1. When the variation wrt F, the deformation gradient is taken, we get the PK1 stress i.e

![neohook4](https://github.com/user-attachments/assets/493bf1cb-b447-446e-8d32-f2780bcdef1d)

Which is used in the stressdivergence kernel, and then when the variation wrt to p is taken, we get 

(J - 1) * del_p

which is implemented to enforce the incompressibility constraint.

The strain energy function was taken from COMSOL documentation for the Neo Hookean model, i.e

![neohook1](https://github.com/user-attachments/assets/bc037bd0-f65c-4cc7-9284-e1570fb6fece)

where:

![neohook2](https://github.com/user-attachments/assets/f3074375-8bee-4025-9813-da73f14acb2c)

and I_1 = trace(C_el)

**Reply by richmondodufisan** [reply]
I also implemented it by numerically differentiating myself (instead of getMaterialPropertyDerivative), and I have slightly better results than before, but it still will not converge beyond higher values. I guess I am now sure that getMaterialPropertyDerivative does not calculate the derivative of a scalar wrt a tensor properly.

The disp_x and y profiles look reasonable, with the y *almost* giving the dog bone shape. But I am still not sure where else the convergence issue could be from. I don't know why none of these are not converging beyond a 1% displacement, whether the MOOSE LinearElastic model or my own custom hyperelastic model.

![disp_x](https://github.com/user-attachments/assets/585de1e3-8d94-471b-95a9-c38ad940ae03)


![disp_y_newtol](https://github.com/user-attachments/assets/b0886d1c-404b-440b-b101-2565d2521773)


test_3.i
https://github.com/richmondodufisan/purple/tree/main/Cornea_Hyperelastic_NeoHookean

**Reply by richmondodufisan** [reply, marked as answer]
Hi, I found the source of the problem. To get it to converge to the right solution I:

-Switched to manual numerical differentiation, was able to get at least one converged timestep. Apparently, the `getMaterialPropertyDerivative` cannot do tensor differentiation.

-Switched to a penalty method. The `PC failed due to FACTOR_NUMERIC_ZEROPIVOT` was coming from the extra field (pressure), and there was no boundary condition to set to it. After implementing that that went away.

-Switched to PK2 formulation (probably didnt do as much to help)

-Set low floating point values to zero, they were coming up in the deformation gradient and seemed to throw the solution in the opposite direction? (see the last result above). I'm not sure how but I remember this being the last thing I did before it worked.


![correct_x](https://github.com/user-attachments/assets/c7528ec2-f1f5-4f2a-b45c-82bb7dc6f0ea)


![correct_y](https://github.com/user-attachments/assets/23426e20-ae89-4521-9046-3486597cf61d)


All this was because I couldn't get a similar problem to converge for MOOSE's linear elastic material for any deformation more than 1%. Maybe using a nearly incompressible poissons ratio on a linear elastic material really is that stiff even with stabilize_strain?
